### ISSUE TITLE
Installation of extra

### ISSUE THREAD

**Issue by AndreaCrotti** [DRMacIver/hypothesis#95]
Maybe is very simple, but I don't get exactly how I should install the packages in extra..

It looks like the version in pypi are older and fail on the requirements (for example hypothesis-datetime), but then if I look at the package installed locally in my virtualenv the extra directory is empty (so they are stripped out).

Why are they just not enabled in the first place and that's it?
What's the point of having different packages in the same repository?

**Comment by DRMacIver** [comment]
You should install them like you install any other package: pip install hypothesis-<package name>. They are not put in the extra directory, they're put in their own package directory like any other package.

Could you elaborate on what you are trying to do that is giving you a version conflict? As far as I can tell all the packages on pypi are up to date.

They are not enabled in the first place and that's it because then I would have to add a bunch of extra dependencies to Hypothesis (I'm aware of the setuptools extra feature, which I'm not currently using as I wasn't when I started this. I may switch over at some point).

What would the point of not having them in the same repository be? How would that materially affect you?

**Comment by AndreaCrotti** [comment]
Ok yes sorry I didn't paste in the error @DRMacIver, I somehow assumed that the versions in Pypi where not the right ones anymore, this is what I get trying to use for example hypothesis-datetime.

```
$~/skimlinks/slx-publisher-stats$ pip install hypothesis-datetime
Collecting hypothesis-datetime
/home/andrea/.virtualenvs/slx-publisher/local/lib/python2.7/site-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:90: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. For more information, see https://urllib3.readthedocs.org/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
Requirement already satisfied (use --upgrade to upgrade): pytz in /home/andrea/.virtualenvs/slx-publisher/lib/python2.7/site-packages (from hypothesis-datetime)
Requirement already satisfied (use --upgrade to upgrade): hypothesis<1.7.99,>=1.7.0 in /home/andrea/.virtualenvs/slx-publisher/lib/python2.7/site-packages (from hypothesis-datetime)
Installing collected packages: hypothesis-datetime
Successfully installed hypothesis-datetime-1.7.1

$~/skimlinks/slx-publisher-stats$ ipython
Python 2.7.6 (default, Mar 22 2014, 22:59:56) 
Type "copyright", "credits" or "license" for more information.

IPython 3.1.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: from hypot
hypothesis          hypothesisdatetime  hypothesispytest    

In [1]: from hypothesis.extra.datetime import datetime
---------------------------------------------------------------------------
VersionConflict                           Traceback (most recent call last)
<ipython-input-1-55d9c75b249b> in <module>()
----> 1 from hypothesis.extra.datetime import datetime

/home/andrea/.virtualenvs/slx-publisher/local/lib/python2.7/site-packages/hypothesis/__init__.py in <module>()
     29 from hypothesis.settings import Settings, Verbosity
     30 from hypothesis.version import __version_info__, __version__
---> 31 from hypothesis.core import given, assume, find, example
     32 
     33 # Force strategy extensions to be loaded here

/home/andrea/.virtualenvs/slx-publisher/local/lib/python2.7/site-packages/hypothesis/core.py in <module>()
    606 
    607 
--> 608 load_entry_points()

/home/andrea/.virtualenvs/slx-publisher/local/lib/python2.7/site-packages/hypothesis/extra/__init__.pyc in load_entry_points(name)
     28         group='hypothesis.extra', name=name
     29     ):
---> 30         package = entry_point.load()  # pragma: no cover
     31         if package not in loaded:
     32             loaded.add(package)

/home/andrea/.virtualenvs/slx-publisher/local/lib/python2.7/site-packages/pkg_resources.pyc in load(self, require, env, installer)
   2045 
   2046     def load(self, require=True, env=None, installer=None):
-> 2047         if require: self.require(env, installer)
   2048         entry = __import__(self.module_name, globals(),globals(), ['__name__'])
   2049         for attr in self.attrs:

/home/andrea/.virtualenvs/slx-publisher/local/lib/python2.7/site-packages/pkg_resources.pyc in require(self, env, installer)
   2058             raise UnknownExtra("Can't require() without a distribution", self)
   2059         list(map(working_set.add,
-> 2060             working_set.resolve(self.dist.requires(self.extras),env,installer)))
   2061 
   2062     @classmethod

/home/andrea/.virtualenvs/slx-publisher/local/lib/python2.7/site-packages/pkg_resources.pyc in resolve(self, requirements, env, installer, replace_conflicting)
    590             if dist not in req:
    591                 # Oops, the "best" so far conflicts with a dependency
--> 592                 raise VersionConflict(dist,req) # XXX put more info here
    593             requirements.extend(dist.requires(req.extras)[::-1])
    594             processed[req] = True

VersionConflict: (hypothesis 1.7.2 (/home/andrea/.virtualenvs/slx-publisher/lib/python2.7/site-packages), Requirement.parse('hypothesis>=1.6,<1.6.99'))
````

**Comment by AndreaCrotti** [comment]
And by the way about not having them installed, there could be also another way, just having them enabled by default and just make it fail if you try to use them without the required dependency.

For example for numpy all you have to check https://github.com/DRMacIver/hypothesis/blob/master/hypothesis-extra/hypothesis-numpy/src/hypothesisnumpy/numpy.py
is that numpy is actually there, but I guess noone would really try to use hypothesis-numpy if it's not in the requirements already so that would not really happen anyway..

**Comment by DRMacIver** [comment]
I think you have a previous version of something installed in your virtualenv. I've literally just tried installing the extra packages in a clean virtualenv (including hypothesis-datetime again just now to double check) and it works fine.

Just assuming packages are installed and failing if they're not is not an adequate solution because it doesn't allow me to express version constraints. Like I said, the correct solution is probably to use setuptools extra_requires, but I didn't know about them when I designed these and haven't had a chance to migrate over yet.

**Comment by AndreaCrotti** [comment]
Mm right I tried with a clean virtualenv and it's true it works, but even after cleaning the current one I can't get it to work, and the error seemed to suggest that hypothesis-datetime wanted an older version than the one installed.
I think anyway I just need to cleanup something so we can close this ticket, thanks

this is the error anyway for future reference:
```
VersionConflict: (hypothesis 1.7.2 (/home/andrea/.virtualenvs/slx-publisher/lib/python2.7/site-packages), Requirement.parse('hypothesis>=1.6,<1.6.99'))
```

**Comment by kxepal** [comment]
Hit the same issue (VersionConflict) with upgrade to 1.8.2.  The problem was fixed with upgrade all installed hypothesis-* packages, even if they were not used directly.

**Comment by DRMacIver** [comment]
Yeah you need to keep installed packages in sync. I'm going to try to fix this for the 0.9 release

**Comment by DRMacIver** [comment]
Err. 1.9 release
